cli/args/mod.rs

```diff
@@ -586,21 +700,68 @@ impl CliOptions {
     .map(Some)
   }
 
+  fn get_npm_process_state(&self) -> Option<NpmProcessState> {
+    if !self.is_npm_main() {
+      return None;
+    }
+
+    let mut maybe_state = self.npm_process_state.lock();
+
+    // TODO(bartlomieju): remove this clone, return a reference maybe?
+    if maybe_state.is_some() {
+      return maybe_state.clone();
+    }
+
+    let state = std::env::var(RESOLUTION_STATE_ENV_VAR_NAME).ok()?;
+    let state: NpmProcessState = serde_json::from_str(&state).ok()?;
+    // remove the environment variable so that sub processes
+    // that are spawned do not also use this.
+    std::env::remove_var(RESOLUTION_STATE_ENV_VAR_NAME);
+    *maybe_state = Some(state.clone());
+    Some(state)
+  }
+
+  pub fn get_npm_resolution_snapshot(&self) -> Option<NpmResolutionSnapshot> {
+    if let Some(state) = self.get_npm_process_state() {
+      return Some(state.snapshot);
+    }
+
+    None
+  }
+
+  // If the main module should be treated as being in an npm package.
+  // This is triggered via a secret environment variable which is used
+  // for functionality like child_process.fork. Users should NOT depend
+  // on this functionality.
+  pub fn is_npm_main(&self) -> bool {
+    *IS_NPM_MAIN
+  }
+
   /// Overrides the import map specifier to use.
   pub fn set_import_map_specifier(&mut self, path: Option<ModuleSpecifier>) {
     self.overrides.import_map_specifier = Some(path);
   }
 
   pub fn node_modules_dir(&self) -> bool {
-    self.flags.node_modules_dir
+    if let Some(node_modules_dir) = self.flags.node_modules_dir {
+      return node_modules_dir;
+    }
+
+    if let Some(npm_process_state) = self.get_npm_process_state() {
```

Comment:
Not ideal this clones a whole npm snapshot potentially, but it's ok for now.

Reply:
Added a TODO to address it.

Reply:
I think you already did address it since it returns a reference now.

Reply:
Yes, but the "snapshot" is still getting cloned when returned.

Reply:
Oh, ok we are talking about different things. This specific call used to clone the whole snapshot, but now it doesn't.